### ISSUE TITLE
git.dirty not working as expected or I haven't understood it properly

### ISSUE THREAD

**Issue by naineel** [ktoso/maven-git-commit-id-plugin#213]
Hi, I've been trying to use the git.dirty feature which according to me should be telling me whether there are modifications in the working directory against the most recent commit. I haven't been able to get it work. Even after there are no differences between the working directory and the most recent commit it still gives me git.dirty=true when I think it should give me false.
I have added the following to the pom.xml: 
```
<plugin>
			<groupId>pl.project13.maven</groupId>
			<artifactId>git-commit-id-plugin</artifactId>
			<version>${plugin.git-commit-id.version}</version>
			<executions>
				<execution>
					<goals>
						<goal>revision</goal>
					</goals>
				</execution>
			</executions>
			<configuration>
				<dateFormat>yyyy.MM.dd '@' HH:mm:ss z</dateFormat>
				<dotGitDirectory>../.git</dotGitDirectory>
				<generateGitPropertiesFile>true</generateGitPropertiesFile>
				<generateGitPropertiesFilename>${project.build.directory}/classes/git.properties</generateGitPropertiesFilename>
				<failOnNoGitDirectory>false</failOnNoGitDirectory>
			</configuration>
		</plugin>
```
Also 
Apache Maven 3.0.5 (r01de14724cdef164cd33c7c8c2fe155faf9602da; 2013-02-19 13:51:28+0000)
Maven home: /usr/share/maven
Java version: 1.8.0_45, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-oracle/jre
Default locale: en_US, platform encoding: ANSI_X3.4-1968
OS name: "linux", version: "3.13.0-57-generic", arch: "amd64", family: "unix"

So basically where am I going wrong ?

**Comment by TheSnoozer** [comment]
Hi @naineel,
thanks for reporting your problem here.
As far as I would say you got the definition of the dirty totally right:
``A working tree is said to be "dirty" if it contains modifications which have not been committed to the current branch.`` (https://www.kernel.org/pub/software/scm/git/docs/gitglossary.html#def_dirty)

What version of the plugin you are currently using?
I just had tested with the latest version available and on my side everything works as expected:

```
$ cd /tmp
$ mvn -version
Apache Maven 3.2.3 (33f8c3e1027c3ddde99d3cdebad2656a31e8fdf4; 2014-08-11T22:58:10+02:00)
Maven home: /usr/share/maven-3.2.3
Java version: 1.8.0_45, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-oracle/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "linux", version: "3.16.0-33-generic", arch: "amd64", family: "unix"
$ git clone git@github.com:ktoso/maven-git-commit-id-plugin.git
$ cd maven-git-commit-id-plugin
$ mvn clean install -Dmaven.test.skip=true && mvn clean package -Pdemo -Dmaven.test.skip=true
$ cat target/testing.properties | grep dirty
git.dirty=false
```

What is the output of ``git status`` when you run the plugin?

Also what does ``git describe --dirty --long`` say?
Does it end with ``-dirty`` (e.g. v2.1.15-16-gf7b1671-dirty)?

**Comment by naineel** [comment]
Hi thanks for the reply. I was using the version 2.1.13 as 14 and 15 aren't compatible with maven 3.0.5.
Git status was showing working directory to be clean. 
git describe --dirty --long was ending with -dirty.

Was this a bug in the 13 version ?

**Comment by TheSnoozer** [comment]
No so far there was not a known bug with the generation of the ``dirty``-flag.

In your case something seems to be strange....when the ``git describe --dirty --long``is ending with ``dirty`` than I would say that the directory is dirty:

https://git-scm.com/docs/git-describe
```
--dirty[=<mark>]
Describe the working tree. It means describe HEAD and appends <mark> (-dirty by default) if the working tree is dirty.
```
Those are native calls when this says it's dirty, it's dirty....

Do you have submodules and maybe something in the submodules is ``dirty``?
What version of git you are using?
Apparently there in one older version of git is a bug (https://stackoverflow.com/questions/16035240/why-is-git-describe-dirty-adding-a-dirty-suffix-when-describing-a-clean-ch)
```
If you are running git 1.7.6 or earlier, you need to run git update-index --refresh before using git describe --dirty, because the index may be stale. 
```

**Comment by naineel** [comment]
I'm using git version 1.7.9.5
The problem is dirty itself works as a command on git. It doesn't reflect on the properties file using the plugin which is the weird part. 
What I forgot to mention is that I'm running this inside a docker container don't think it matters though.
Do you have any other ideas as to why it's happening ?
Any other details I can give you ?

**Comment by TheSnoozer** [comment]
Mhh this is kind of strange. I could try to trace this issue down. The docker shouldn't do any difference.
Can you do me the favour and set the additional configuration ``<useNativeGit>true</useNativeGit>`` for the plugin? 
This will execute the queries we gather inside the plugin directly to the native git binary (without the config it will ask a custom jgit implementation). If it produces the same results I would doubt that we can do something here.

And also to clarify we more or less have two different types of the ``dirty`` the one that is supposed to be appended to the ``git.describe`` and the other one supposed to indicate that the repro is ``dirty``...we both are talking about the property that indicates that the repro is ``dirty`` right? :-)

**Comment by naineel** [comment]
Well I actually tried that after reading the documentation when it wasn't working and it produced the same result. :( 
I actually looked at the source code and especially the tests which checked the dirty functionality. But just wasn't able to reproduce it.
And yeah we were pretty much talking about the same dirty.
Thanks for all the help..I guess you can close the issue now..I already used an alternate solution which involved writing a bash script.

**Comment by TheSnoozer** [comment]
Well I would love to fix it when it's a problem or bug and I would give it a shot to debug. 
However when the native git binary says it's dirty I also would say it's dirty :-)
When you run the plugin with the native binary we basically execute ``git status -s`` and check if the result is empty. So maybe in some strange version of git this returns a newline or whatsoever....

What docker container are you using (a public one or custom tailored)?

Does the alternative solution (with writing bash scripts) what are you invoking to check whether a repro is dirty or not? (maybe we are just using a wrong way to check for dirty and could deploy a better way to do so...)

**Comment by naineel** [comment]
Its a custom tailored docker container though I doubt that's the issue.
Well I used git dirty --exit-code and checked $? 
What this does is it essentially reduces the output to a 1 or a 0.. 1 if dirty and 0 if clean. 
Use it if it is feasible :)

**Comment by TheSnoozer** [comment]
Where did you get the git binaries from?
Apparently those from https://git-scm.com/docs does not support the commands ``dirty`` and ``--exit-code``

Is this the binary from https://code.google.com/p/git-core/ ?

**Comment by naineel** [comment]
Well it is there in the git-scm..check it out 
http://git-scm.com/docs/git-diff

**Comment by naineel** [comment]
Oops Im so sorry..I made a typo in the command I sent you..its actually
git diff --exit-code not dirty

**Comment by TheSnoozer** [comment]
Using: https://registry.hub.docker.com/_/ubuntu/

Dockerfile:
```
# Pull base image.
# FROM dockerfile/ubuntu
FROM ubuntu:14.04.1

# Install Java, git and maven
RUN \
  apt-get update && \
  apt-get -y --ignore-missing install python-software-properties && \
  apt-get -y --ignore-missing install software-properties-common && \
  add-apt-repository -y ppa:webupd8team/java && \
  apt-get update && \
  apt-get -y install git-core maven && \
  echo debconf shared/accepted-oracle-license-v1-1 select true | sudo debconf-set-selections && \
  echo debconf shared/accepted-oracle-license-v1-1 seen true | sudo debconf-set-selections && \
  apt-get -y install oracle-java8-installer && \
  rm -rf /var/lib/apt/lists/*

# Define working directory.
WORKDIR /tmp

# Define commonly used JAVA_HOME variable
ENV JAVA_HOME $(echo $(readlink $(readlink $(which java))) | sed 's:/bin/java::')

# Define default command.
CMD ["bash"]
```

Steps to build ubuntu docker container (draft):
``
$ cd folder/where/dockerfile/is/located/
$ docker build -t thesnoozer/maven-git-commit-id-plugin .
$ docker run -it thesnoozer/maven-git-commit-id-plugin
# cd /tmp
# git --version
# mvn -version
# git clone git@github.com:ktoso/maven-git-commit-id-plugin.git
# cd maven-git-commit-id-plugin
# mvn clean install -Dmaven.test.skip=true && mvn clean package -Pdemo -Dmaven.test.skip=true
# cat target/testing.properties | grep dirty
```

**Comment by TheSnoozer** [comment]
Have tested with docker ubuntu:14.04.2 and ubuntu:12.04.5 used different maven versions and different git versions. However the dirty-flag in v2.1.13 works as expected. 

If someone else comes across this issue I would love to investigate but I think at this stage I would say everything works as expected.
